File: src/main/java/no/novari/flyt/archive/gateway/template/validation/UniqueKeys.java
package no.novari.flyt.archive.gateway.template.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = UniqueKeysValidator.class)
public @interface UniqueKeys {

    String DUPLICATE_KEYS_REF = "duplicateKeys";

    String message() default "contains duplicate keys: {" + DUPLICATE_KEYS_REF + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/template/validation/UniqueKeys.kt
package no.novari.flyt.archive.gateway.template.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [UniqueKeysValidator::class])
annotation class UniqueKeys(
    val message: String = "contains duplicate keys: {$DUPLICATE_KEYS_REF}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val DUPLICATE_KEYS_REF = "duplicateKeys"
    }
}
